File: content/examples/other/leetcode/problems/same-tree/example-2/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function isSameTreeIterative(p, q) {
  const stack = [[p, q]];
  while (stack.length > 0) {
    const [a, b] = stack.pop();
    if (a == null && b == null) continue;
    if (a == null || b == null) return false;
    if (a.val !== b.val) return false;
    stack.push([a.left, b.left], [a.right, b.right]);
  }
  return true;
}

if (require.main === module) {
  const p = new TreeNode(1, new TreeNode(2), new TreeNode(3));
  const q = new TreeNode(1, new TreeNode(2), new TreeNode(3));
  console.log(isSameTreeIterative(p, q));
}

module.exports = { TreeNode, isSameTreeIterative };

File: content/examples/other/leetcode/problems/same-tree/example-2/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def is_same_tree_iterative(p, q):
    stack = [(p, q)]
    while stack:
        a, b = stack.pop()
        if a is None and b is None:
            continue
        if a is None or b is None:
            return False
        if a.val != b.val:
            return False
        stack.append((a.left, b.left))
        stack.append((a.right, b.right))
    return True


if __name__ == "__main__":
    p = TreeNode(1, TreeNode(2), TreeNode(3))
    q = TreeNode(1, TreeNode(2), TreeNode(3))
    print(is_same_tree_iterative(p, q))
